oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/SpringOpaqueTokenIntrospector.java

```diff
@@ -295,4 +311,87 @@ default List<String> getScopes() {
 
 	}
 
+	/**
+	 * Used to build {@link SpringOpaqueTokenIntrospector}.
+	 *
+	 * @author Ngoc Nhan
+	 * @since 6.5
+	 */
+	public static final class Builder {
+
+		private final String introspectionUri;
+
+		private String clientId;
+
+		private String clientSecret;
+
+		private Builder(String introspectionUri) {
+			this.introspectionUri = introspectionUri;
+		}
+
+		/**
+		 * Uses the given parameters to build {@code SpringOpaqueTokenIntrospector}
+		 * @param clientId The client id authorized that should be encoded
+		 * @param charset The charset to use
+		 * @return the {@link SpringOpaqueTokenIntrospector.Builder}
+		 * @since 6.5
+		 */
+		public Builder clientId(String clientId, Charset charset) {
+			Assert.notNull(clientId, "clientId cannot be null");
+			Assert.notNull(charset, "charset cannot be null");
+			this.clientId = URLEncoder.encode(clientId, charset);
+			return this;
+		}
+
+		/**
+		 * Uses the given parameter to build {@code SpringOpaqueTokenIntrospector}
+		 * @param clientId The client id authorized
+		 * @return the {@link SpringOpaqueTokenIntrospector.Builder}
+		 * @since 6.5
+		 */
+		public Builder clientId(String clientId) {
```

Comment:
While I see the value of having both encoded and unencoded methods here, I think that distinction will be lost on most users of the API. Let's please only have `clientId`, which takes an unencoded value and encodes it using `UTF_8` and `clientSecret`, which takes an unencoded value and encodes it using `UTF_8`. The builder should then only have three public methods.

If folks need a different `CharSet` or if they need to encode the values themselves, they should use the `RestOperations` constructor (or the deprecated constructor for the time being).